### PR TITLE
Put the enc config in a separate class 

### PR DESCRIPTION
**Pull request by SimonPe**
This makes it possible to "autodiscover" the enc using exported
resources.

This sits on top of the reworked #364

### CODE REVIEW

**Comment by mmoll** [comment]
please rebase :)

**Comment by SimonPe** [comment]
@mmoll here you are.
sorry it took so long, I was a bit busy with other parts of our infra.

**Comment by SimonPe** [comment]
@mmoll ping?

**Comment by mmoll** [comment]
Sorry, I don't have much time at the moment, @ekohl @domcleal, what do you think? It's looking fine for me in general, however, I don't know if adding all the parameters to everything is a big downside.

**Comment by SimonPe** [comment]
@mmoll  @domcleal @ekohl any update on this?

**Comment by mmoll** [comment]
I'd like to bring this in generally, reviewed each change some weeks ago and think if tests are green after a rebase this is good to go.

@ekohl @domcleal any objections?

Also see https://github.com/theforeman/puppet-puppet/pull/364#issuecomment-213613886

**Comment by domcleal** [comment]
@mmoll I've not reviewed it in detail, but I think I'm OK with the principle of the change, if you did want to merge it. It looks more reliable to manage the file entries from the manifest instead of keeping logic in the template.

File: manifests/config/additional_settings.pp [hunk 1]
```diff
@@ -0,0 +1,12 @@
+define puppet::config::additional_settings(
+  $hash,
+  $resource,
+  $key = $title,
+  $options = {},
+){
+  create_resources($resource, {
+      "${key}" => {
```
Comment (ekohl) [inline manifests/config/additional_settings.pp]: Is this needed or can it just be ```$key => {```?

File: manifests/config/entry.pp [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+  # note the spaces at he end of the 'order' parameters,
+  # they make sure that '1_main ' is ordered before '1_main_*'
+  ensure_resource('concat::fragment', "puppet.conf_${section}", {
+    target  => "${::puppet::dir}/puppet.conf",
+    content => "\n\n[${section}]",
+    order   => "${sectionorder}_${section} ",
+  })
+
+  if (!defined(Concat::Fragment["puppet.conf_${section}_${key}"])){
```
Comment (ekohl) [inline manifests/config/entry.pp]: This could use a comment that it appends the value when the value is used multiple times.

Comment (SimonPe) [inline manifests/config/entry.pp]: added a comment now.

**Comment by SimonPe** [comment]
@ekohl @mmoll : thanks for your patience, i rebased it now and added the comment @ekohl asked for, the "${key}" is still in there since the old parser requires this.

**Comment by mmoll** [comment]
This is looking good and survived testing in my (small) environment. @ekohl what's your opinion?

**Comment by carroarmato0** [comment]
This would be lovely

**Comment by ekohl** [comment]
I haven't tested it, but the code looks good. Given @mmoll did some testing and we all agree it's a good idea I think we should merge this.

**Comment by mmoll** [comment]
cherry-picked as

commit  | message
------------ | -------------
06dfd8c | Put the enc config in a separate class
bfbc974 | Add backwards compatible additional_settings handling
51a3570 | use resources to manage puppet.conf
0d25365 | add puppet::config::* resource definitions


hartelijk dank aan @SimonPe!

**Comment by ekohl** [comment]
Big thank you to @SimonPe, this was a big refactor and thank you for your patience to stick with it.

**Pull request closed without merging**